Replacement implementation for groupslices and associated functions

AndyGreenwell opened the pull request:
This PR replaces the original functionality proposed in issue #14142.

The main groupslices function returns the vector `uniquerow` (also referred to as `ic` in the prior discussion) as originally calculated by the `unique(A,dim)` implementation in multidimensional.jl.  A series of wrapper functions (`groupinds`, `firstinds`, `lastinds`) are included for returning the vector `ia` and vector of vectors `ib`, as discussed within #14142.

AndyGreenwell committed:
Adding updated API signatures based on discussion in #14142

This commit adds updated argument signatures for the functionality discussed in #14142 wherein each of these functions only returns one single output argument containing index vectors of various purposes related to the unique slices currently returned from the unique function taking two input arguments.  Still needs tests and doc integration.

AndyGreenwell committed:
Adding updated API signatures based on discussion in #14142

This commit adds updated argument signatures for the functionality discussed in #14142 wherein each of these functions only returns one single output argument containing index vectors of various purposes related to the unique slices currently returned from the unique function taking two input arguments.  Still needs tests and doc integration.

AndyGreenwell committed:
Adding updated API signatures based on discussion in #14142

This commit adds updated argument signatures for the functionality discussed in #14142 wherein each of these functions only returns one single output argument containing index vectors of various purposes related to the unique slices currently returned from the unique function taking two input arguments.  Still needs tests and doc integration.

AndyGreenwell committed:
uniquerow is ic, and seems to handle hash collisions already

This commit alters the current groupslices function to return the vector uniquerow that was originally calculated within the existing unique function.  The values contained within uniquerow for cases where there are no hash collisions are actually equal to what I was calculating in array ic.  As @simonster pointed out in comment https://github.com/JuliaLang/julia/pull/14142#issuecomment-193089681 the previous commit was not taking into account hash collisions for the values in ic.  As uniquerow within unique was already calculating the values in ic, taking into account hash collisons, and updating its values accordingly, we can just return uniquerow from groupslices.  For continuity with the conversation in #14142, I currently have assigned ic as an alias for uniquerow, but that can certainly be removed.

Change error to ArgumentError

uniquerow is ic, and seems to handle hash collisions already

Add initial tests for multidimensional.jl for groupslices and related functions

This commit adds some initial tests for the functions groupslices, groupinds, firstinds, and lastinds.  The initial tests are very basic and do not currently include a test for a multidimensional input array that would have hash collisions between slices.

AndyGreenwell committed:
uniquerow is ic, and seems to handle hash collisions already

This commit alters the current groupslices function to return the vector uniquerow that was originally calculated within the existing unique function.  The values contained within uniquerow for cases where there are no hash collisions are actually equal to what I was calculating in array ic.  As @simonster pointed out in comment https://github.com/JuliaLang/julia/pull/14142#issuecomment-193089681 the previous commit was not taking into account hash collisions for the values in ic.  As uniquerow within unique was already calculating the values in ic, taking into account hash collisons, and updating its values accordingly, we can just return uniquerow from groupslices.  For continuity with the conversation in #14142, I currently have assigned ic as an alias for uniquerow, but that can certainly be removed.

AndyGreenwell committed:
uniquerow is ic, and seems to handle hash collisions already

This commit alters the current groupslices function to return the vector uniquerow that was originally calculated within the existing unique function.  The values contained within uniquerow for cases where there are no hash collisions are actually equal to what I was calculating in array ic.  As @simonster pointed out in comment https://github.com/JuliaLang/julia/pull/14142#issuecomment-193089681 the previous commit was not taking into account hash collisions for the values in ic.  As uniquerow within unique was already calculating the values in ic, taking into account hash collisons, and updating its values accordingly, we can just return uniquerow from groupslices.  For continuity with the conversation in #14142, I currently have assigned ic as an alias for uniquerow, but that can certainly be removed.

AndyGreenwell committed:
Change error to ArgumentError

AndyGreenwell committed:
Change error to ArgumentError

AndyGreenwell committed:
Merge branch 'master' of https://github.com/AndyGreenwell/julia

AndyGreenwell committed:
Add initial tests for multidimensional.jl for groupslices and related functions

This commit adds some initial tests for the functions groupslices, groupinds, firstinds, and lastinds.  The initial tests are very basic and do not currently include a test for a multidimensional input array that would have hash collisions between slices.

tkelman commented on test/multidimensional.jl:
```diff
@@ -0,0 +1,30 @@
+# This file is a part of Julia. License is MIT: http://julialang.org/license
+
+using Base.Test
```
need to add this file to the list in `test/choosetests.jl` for it to run

AndyGreenwell committed:
Remove whitespace, add entry to choosetests.jl

This commit addresses the removal of trailing whitespace from prior commit, and also adds an entry for multidimensional to test/choosetests.jl

AndyGreenwell committed:
Add exports for groupslices and associated functions

AndyGreenwell committed:
Add a space, trigger test execution

AndyGreenwell committed:
Fix merge conflicts

AndyGreenwell committed:
pull upstream master and fix merge conflicts

JoshChristie commented on base/multidimensional.jl:
```diff
@@ -931,3 +931,171 @@ If `dim` is specified, returns unique regions of the array `itr` along `dim`.
+that are present in the array `C` as returned from `unique(A, dim)`.
+"""
+
+function groupinds(ic::Vector{Int})
```
This doesn't seem to work in my use case. (I haven't switched to your branch or updated to 0.5. I just copied your code changes)
```julia
A =  [1.0  1.0  2.0  2.0  1.0  1.0  1.0  2.0  2.0  2.0
      1.0  1.0  2.0  2.0  2.0  1.0  2.0  2.0  2.0  1.0
      1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  2.0]
ic = groupslices(A, 2)
ib = groupinds(ic)
ERROR: BoundsError: attempt to access 4-element Array{Array{Int64,1},1}:
 [1,2]  
 Int64[]
 [3,4]  
 Int64[]
  at index [5]
in groupinds at 1046
```

My fix: 
```julia
function groupinds(ic::Vector{Int})
    d = Dict{Int, Int}()
    ia = unique(ic)
    n = length(ia)
    for i = 1:n
        d[ia[i]]= i
    end

    ib = Array(Vector{Int},n)
    for k = 1:n
        ib[k] = Int[]
    end

    for h = 1:length(ic)
        push!(ib[d[ic[i]]], h)
    end
    return ib
end
```

AndyGreenwell replied:
Thanks, using a Dict similar to `d` above is along the lines of what was in the original proposal [here](https://github.com/JuliaLang/julia/pull/14142/files#diff-89e97f67f8058d32eac8d917db5bbdbfR977) and [here](https://github.com/JuliaLang/julia/pull/14142/files#diff-89e97f67f8058d32eac8d917db5bbdbfR1045).  I'll incorporate your changes.

AndyGreenwell committed:
Addressing line note comment for groupinds

This commit incorporates the fix for groupinds provided by @JoshChristie, and adds the provided test case from that line note to test/multidimensional.jl

tkelman commented on base/multidimensional.jl:
```diff
@@ -940,3 +940,178 @@ If `dim` is specified, returns unique regions of the array `itr` along `dim`.
+   all(A .== C[:,:,ic])
+end
+"""
+
```
remove the empty line between the docstring and the function it is documenting

tkelman commented on base/multidimensional.jl:
```diff
@@ -940,3 +940,178 @@ If `dim` is specified, returns unique regions of the array `itr` along `dim`.
 end
+
+"""
+    groupslices(A, dim)
```
see CONTRIBUTING.md, the signature line needs to be added to the rst docs, then run `make docs` to populate the docstring content

AndyGreenwell replied:
In which *.rst file should the signature line for this function live?  At present, there is no multidimensional.rst file.

tkelman commented on base/multidimensional.jl:
```diff
@@ -940,3 +940,178 @@ If `dim` is specified, returns unique regions of the array `itr` along `dim`.
+Example usage:
+
+If `C = unique(A, dim)`, `ic = groupslices(A, dim)`, and
+rank(A) == rank(C) == 3, then:
```
should have backticks, and I think you mean `ndims` here not linear algebra rank

AndyGreenwell replied:
Will update.

tkelman commented on base/multidimensional.jl:
```diff
@@ -940,3 +940,178 @@ If `dim` is specified, returns unique regions of the array `itr` along `dim`.
+If `C = unique(A, dim)`, `ic = groupslices(A, dim)`, and
+rank(A) == rank(C) == 3, then:
+
+if dim == 1
```
code examples should be indented so the doc systen renders them as such

tkelman commented on base/multidimensional.jl:
```diff
@@ -940,3 +940,178 @@ If `dim` is specified, returns unique regions of the array `itr` along `dim`.
+grouped vector of vectors.  Each vector entry in the returned vector of
+vectors contains all of the positional indices of slices in the original
+input array `A` that correspond to the unique slices along dimension `dim`
+that are present in the array `C` as returned from `unique(A, dim)`.
```
the wording here is tough to follow - should probably be rephrased

JoshChristie commented on base/multidimensional.jl:
```diff
@@ -940,3 +940,178 @@ If `dim` is specified, returns unique regions of the array `itr` along `dim`.
+   all(A .== C[:,ic,:])
+elseif dim == 3
+   all(A .== C[:,:,ic])
+end
```
This example usage fails for me. 

```julia
A =  [1.0  1.0  2.0  2.0  1.0  1.0  1.0  2.0  2.0  2.0
      1.0  1.0  2.0  2.0  2.0  1.0  2.0  2.0  2.0  1.0
      1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  2.0]
dim = 2
C = unique(A, dim)
ic = groupslices(A, dim)

if dim == 1
   all(A .== C[ic,:,:])
elseif dim == 2
   all(A .== C[:,ic,:])
elseif dim == 3
   all(A .== C[:,:,ic])
end

ERROR: BoundsError: attempt to access 3x4 Array{Float64,2}:
 1.0  2.0  1.0  2.0
 1.0  2.0  2.0  1.0
 1.0  1.0  1.0  2.0
  at index [Colon(),[1,1,3,3,5,1,5,3,3,10],Colon()]
 in throw_boundserror at abstractarray.jl:156
 in _internal_checkbounds at abstractarray.jl:176
 in checkbounds at abstractarray.jl:159
 in _getindex at multidimensional.jl:185
 in getindex at abstractarray.jl:488
```

AndyGreenwell replied:
Before I make another commit, I need to work get the documentation strings integrated properly as @tkelman points out, but the changes in this [gist](https://gist.github.com/AndyGreenwell/4c2f1481f5c974eb14f8c8266affe85b) should allow your example to work appropriately.  In one of the changes made in the course of discussing #14142 related to detecting hash collisions, where `ic` being directly assigned the values of `uniquerow` was improper.  While the values of `uniquerow` do contain the indices of unique slices of `A`, they are not the indices corresponding to the unique slices in `C` that allow for reconstruction of `A`.

The changes in this [gist](https://gist.github.com/AndyGreenwell/4c2f1481f5c974eb14f8c8266affe85b) assign the appropriate values to `ic` that allow one to reconstruct `A` by indexing `C` with `ic`, as had been the case in the original `uniqueslices` submission (but had not been handling hash collisions properly).

Thank you for testing.

AndyGreenwell commented:
The functionality from this PR is now included in the following repository:

https://github.com/AndyGreenwell/GroupSlices.jl

That repo has been tagged, registered, and submitted as a [PR](https://github.com/JuliaLang/METADATA.jl/pull/5726) to METADATA.jl.  Once that PR is merged in METADATA, then I will close this PR.
